File: Lab2/haskell/TypeChecker.hs
{-# LANGUAGE FlexibleContexts #-}
{-# LANGUAGE LambdaCase #-}

module TypeChecker where

import Control.Monad

import Data.Map (Map)
import qualified Data.Map as Map
import qualified Data.List as List


import Data.Maybe

import CPP.Abs
import CPP.Print
import CPP.ErrM

type Env = (Sig, [Context]) -- functions and context stack
type Sig = Map Id ([Type], Type) -- function type signature
type Context = Map Id Type -- variables with their types

lookupVar :: Env -> Id -> Err Type
lookupVar (_sig,[]) x = fail $ "Unknown variable " ++ printTree x ++ "."
lookupVar (sig, scope:rest) x = case Map.lookup x scope of
    Nothing -> lookupVar (sig, rest) x
    Just t  -> return t

lookupFun :: Env -> Id -> Err ([Type],Type)
lookupFun (sig, _) x = case Map.lookup x sig of
    Nothing -> fail $ "Function " ++ printTree x ++ " is not declared."
    Just t  -> return t

{-updateVar :: Env -> Id -> Type -> Err Env
updateVar (sig, scope:rest) x t = case Map.lookup x scope of
    Nothing -> updateVar (sig, rest) x t
    Just t1 -> do
        when (t1 /= t) $
            fail $ "Type of " ++ printTree x ++ " does not match given type"
        return (sig, (Map.insert x t scope):rest)
updateVar (_,[]) x _ = fail $ ("Variable " ++ printTree x
                             ++ " is not declared.") -}
                             
updateVar :: Env -> Id -> Type -> Err Env
updateVar (sig,cxt) id typ = do
  when (typ==Type_void) $ fail "type void is illegal for variable"
  case cxt of
    [] -> fail "type checker: no context"
    c:cs -> case Map.lookup id c of
      Just a -> fail $ "variable " ++ printTree id ++ " declared already"
      Nothing -> return (sig, Map.insert id typ c :cs)

newVar :: Env -> Id -> Type -> Err Env
newVar (sig, scope:rest) x t = case Map.lookup x scope of
    Nothing -> do
        when (t == Type_void) $
            fail $ "Type void of variable " ++ printTree x ++ " is illegal"
        return (sig, (Map.insert x t scope):rest)
    Just _  -> fail $"Variable " ++ printTree x ++ " already declared."

updateFun :: Env -> Id -> ([Type],Type) -> Err Env
updateFun (sig, ctx) x t = case Map.lookup x sig of
    Nothing -> return ((Map.insert x t sig), ctx)
    Just _ -> fail $ "Function " ++ printTree x ++ " is already declared."

updateDefs :: Program -> Err Sig
updateDefs (PDefs ds) = do
  -- add function definitions to 
  let fs = map getDefType ds ++ builtin
      s  = Map.fromList fs
  -- check for duplicate function id:s
  let names = map fst fs
      dup = names List.\\ List.nub names
  unless (null dup) $ fail $
    "the following functions are defined several times: " ++
    List.intercalate ", " (map printTree dup)
  return s
  where getDefType (DFun t id as _) = (id, (map getArgType as,t))
        getArgType (ADecl t id)     =  t
        
newBlock :: Env -> Env
--newBlock (sig, ctx) = (sig, [])
newBlock (sig,cs) = (sig,Map.fromList [] : cs)

emptyEnv :: Env
emptyEnv = (Map.empty, [])

{-inferExp :: Env -> Exp -> Err Type
inferExp env x = case x of
    ETrue   -> return Type_bool
    EInt _  -> return Type_int
    EDouble _ -> return Type_double
    EId id  -> lookupVar env id
    EPlus exp1 exp2 ->
        inferBin[Type_int, Type_double] env exp1 exp2
    EMinus exp1 exp2 ->
        inferBin[Type_int, Type_double] env exp1 exp2
    EDiv exp1 exp2 ->
        inferBin[Type_int, Type_double] env exp1 exp2
    ETimes exp1 exp2 ->
        inferBin[Type_int, Type_double] env exp1 exp2
    EApp id exps -> do 
        (args, typ) <- lookupFun env id
        exps'     <- mapM (inferExp env) exps
        if (args == exps')
            then return typ
        else fail $ "The arguments (" ++ printTree exps' 
            ++ ") don't match parameters (" ++ printTree args ++ ")"
    EPostIncr id -> do
        typ <- lookupVar env id
        inferUn [Type_int, Type_double] env typ
    EPreIncr id -> do
        typ <- lookupVar env id
        inferUn [Type_int, Type_double] env typ
    EPostDecr id -> do
        typ <- lookupVar env id
        inferUn [Type_int, Type_double] env typ
    EPreDecr id -> do
        typ <- lookupVar env id
        inferUn [Type_int, Type_double] env typ
    ELt exp1 exp2 ->
        inferBin [Type_int, Type_double] env exp1 exp2
    EGt exp1 exp2 ->
        inferBin [Type_int, Type_double] env exp1 exp2
    ELtEq exp1 exp2 ->
        inferBin [Type_int, Type_double] env exp1 exp2
    EGtEq exp1 exp2 ->
        inferBin [Type_int, Type_double] env exp1 exp2
    EEq exp1 exp2 ->
        inferBin [Type_int, Type_double, Type_bool] env exp1 exp2
    ENEq exp1 exp2 ->
        inferBin [Type_int, Type_double, Type_bool] env exp1 exp2
    EAnd exp1 exp2 ->
        inferBin [Type_bool] env exp1 exp2
    EOr exp1 exp2 ->
        inferBin [Type_bool] env exp1 exp2
    EAss id exp -> do
 {-       typ <- lookupVar env id
        inferAss [Type_double, Type_bool, Type_int] env typ exp-}
       typ <- lookupVar env id
       checkExp env typ exp
       return typ-}

inferExp :: Env -> Exp -> Err Type
inferExp env exp = case exp of
  ETrue     -> return Type_bool
  EFalse    -> return Type_bool
  EInt _    -> return Type_int
  EDouble _ -> return Type_double
  EId id    -> lookupVar env id
  EApp id es ->
    do (typs,typ) <- lookupFun env id
       args <- mapM (inferExp env) es
       unless (typs == args)
         $ fail $ "types of: " ++
           List.intercalate ", " (map printTree typs) ++
           "\nexpected for function " ++ printTree id ++ 
           "\nbut found: " ++ List.intercalate ", " (map printTree args)
       return typ
  EPostIncr id -> incdec id
  EPostDecr id -> incdec id
  EPreIncr id  -> incdec id
  EPreDecr id  -> incdec id
  ETimes e1 e2 -> inferBin [Type_int,Type_double] env e1 e2
  EDiv   e1 e2 -> inferBin [Type_int,Type_double] env e1 e2
  EPlus  e1 e2 -> inferBin [Type_int,Type_double] env e1 e2
  EMinus e1 e2 -> inferBin [Type_int,Type_double] env e1 e2
  ELt    e1 e2 -> inferBin [Type_int,Type_double,Type_bool] env e1 e2 >>
                           return Type_bool
  EGt    e1 e2 -> inferBin [Type_int,Type_double,Type_bool] env e1 e2 >>
                           return Type_bool
  ELtEq  e1 e2 -> inferBin [Type_int,Type_double,Type_bool] env e1 e2 >>
                           return Type_bool
  EGtEq  e1 e2 -> inferBin [Type_int,Type_double,Type_bool] env e1 e2 >>
                           return Type_bool
  EAnd   e1 e2 -> inferBin [Type_bool] env e1 e2
  EOr    e1 e2 -> inferBin [Type_bool] env e1 e2
  EEq    e1 e2 -> inferBin [Type_int,Type_double,Type_bool] env e1 e2 >>
                           return Type_bool
  ENEq   e1 e2 -> inferBin [Type_int,Type_double,Type_bool] env e1 e2 >>
                           return Type_bool
  EAss id exp -> do
    typ <- lookupVar env id
    checkExp env typ exp
    return typ  
  where incdec id =
          do typ <- lookupVar env id
             unless (typ == Type_int || typ == Type_double)
               $ fail $ "type of int or double expected but found " ++
                        printTree typ
             return typ


inferBin :: [Type] -> Env -> Exp -> Exp -> Err Type
{-inferBin types env exp1 exp2 = do
    typ <- inferExp env exp1
    if elem typ types
        then do
            checkExp env typ exp2
            return typ

    else
        fail $ "wrong type of expression " ++ printTree exp1-}

inferBin typs env exp1 exp2 = do
  typ <- inferExp env exp1
  unless (elem typ typs) $ fail $ "wrong type of expression " ++ printTree exp1
  checkExp env typ exp2
  return typ

inferAss :: [Type] -> Env -> Type -> Exp -> Err Type
inferAss types env typ exp = do
    if elem typ types
        then do
            checkExp env typ exp
            return typ

    else
        fail $ "wrong type of expression " ++ printTree exp

inferUn :: [Type] -> Env -> Type -> Err Type
inferUn types env typ = do
    if elem typ types
        then
            return typ
    else
        fail $ "Type " ++ show typ ++ " does not match"


checkExp :: Env -> Type -> Exp -> Err ()
checkExp env typ exp = do
    typ2 <- inferExp env exp
    if (typ2 == typ) then
        return ()
    else
        fail $ "Type of " ++ printTree exp ++
        "expected " ++ printTree typ ++
        "but found " ++ printTree typ2


checkStm :: Env -> Type -> Stm -> Err Env
checkStm env val x = case x of
    SExp exp -> do
        inferExp env exp
        return env
 {-   SDecls typ xs -> do
        case xs of
            x:xs  -> do 
                env' <- updateVar env x typ
                checkStm env' typ (SDecls typ xs)
            []    -> fail $ "No more variables" ++ show typ -}
    SDecls typ x ->     
        foldM (\env id -> updateVar env id typ) env x
            
    SWhile exp stm -> do
        checkExp env Type_bool exp
        checkStm env val stm
    SInit typ id exp -> do
        expTyp <- inferExp env exp
        if (typ == expTyp) then
            updateVar env id typ
        else
            fail $ "Type " ++ show typ ++ "does not match"
                ++ printTree exp
    SBlock stms -> do
        checkStms (newBlock env) val stms
        return env
        
    SReturn exp -> do
        typ <- inferExp env exp
        if typ /= val then
            fail $ "Type " ++ show typ ++ "does not match " ++
            "expected type" ++ show val
        else
            return env
    SIfElse exp stm1 stm2 -> do
 {-       checkExp env Type_bool exp
        checkStm env val stm1
        checkStm env val stm2 -}
                 checkExp env Type_bool exp
                 checkStm (newBlock env) val stm1
                 checkStm (newBlock env) val stm2
                 return env

checkStms :: Env -> Type -> [Stm] -> Err Env
checkStms env typ = foldM (\env stm -> checkStm env typ stm) env 
{-checkStms env typ stms = case stms of
    stm:stms -> do env' <- checkStm env typ stm
                   checkStms env' typ stms
    []       -> return env]-}

typecheck :: Program -> Err ()
typecheck (PDefs defs) = do
    env <- foldM (\env (id, (typ)) -> updateFun env id typ) emptyEnv builtin
    env' <- foldM (\env' (DFun typ id args _) -> updateFun env'
                    id (handleArgs args, typ)) env defs
    mapM_ (checkDef env') defs


{-checkDef :: Env -> Def -> Err ()
checkDef env (DFun t _ args stms) = do
    mapM_ (\ (ADecl t x) -> updateVar env x t) args
    mapM_ (checkStm env t) stms-}

checkArgs :: Env -> [Arg] -> Err Env
checkArgs = foldM updateArgs
  where updateArgs env arg = updateVar env id typ
          where (ADecl typ id) = arg


checkDef :: Env -> Def -> Err ()
checkDef env (DFun typ id args stms) = do
  env' <- checkArgs (newBlock env) args
  checkStms env' typ stms
  return ()

checkDefs :: Env -> [Def] -> Err ()
checkDefs env = foldM (\_ def -> checkDef env def) ()


builtin :: [(Id, ([Type],Type))]
builtin =
    [ (Id "readInt"    , ([], Type_int))
    , (Id "readDouble" , ([], Type_double))
    , (Id "printInt"   , ([Type_int], Type_void ))
    , (Id "printDouble", ([Type_double], Type_void))
    ]
handleArgs :: [Arg] -> [Type]
handleArgs xs = do
    tList <- map (\(ADecl typ id) -> typ) xs
    return tList

nyi = error "NOT YET IMPLEMENTED"

File: Lab2/java/TypeChecker.java
import CPP.Absyn.*;
import java.util.*;

public class TypeChecker {
    // Global signature of functions
    private Map<String, FunType> sig;

    // Stack of contexts
    private List<Map<String,Type>> cxt;

    // Return type of function we are checking
    private Type returnType;

    // Share type constants
    public final Type BOOL   = new Type_bool();
    public final Type INT    = new Type_int();
    public final Type DOUBLE = new Type_double();
    public final Type VOID   = new Type_void();

    // Entry point
    public void typecheck(Program p) {
        p.accept(new ProgramVisitor(), null);
    }

    ////////////////////////////// Program //////////////////////////////

    public class ProgramVisitor implements Program.Visitor<Void,Void> {
        public Void visit(CPP.Absyn.PDefs p, Void arg) {
            // Put primitive functions into signature
            sig = new TreeMap<String, FunType>();
            sig.put("printInt"   , new FunType(VOID, singleArg(INT)));
            sig.put("readInt"    , new FunType(INT, new ListArg()));
            sig.put("printDouble", new FunType(VOID, singleArg(DOUBLE)));
            sig.put("readDouble" , new FunType(DOUBLE, new ListArg()));

            // Extend signature by all the definitions
            for (Def d: p.listdef_) {
                d.accept(new DefIntoSigVisitor(), null);
            }

            // Check definitions
            for (Def d: p.listdef_) {
                d.accept(new DefVisitor(), arg);
            }

            // Check for main
            if(!sig.containsKey("main"))
                throw new TypeException("The 'main' function is missing!");
            if(!sig.get("main").returnType.equals(INT))
			        	throw new TypeException("The 'main' function has wrong return type");
		      	if(!sig.get("main").args.isEmpty())
				        throw new TypeException("The 'main' function has no argument");
            return null;
        }
    }

    public ListArg singleArg (Type t) {
        ListArg l = new ListArg();
        l.add(new ADecl(t, "x"));
        return l;
    }

    ////////////////////////////// Function //////////////////////////////

    public class DefIntoSigVisitor implements Def.Visitor<Void,Void> {
        public Void visit(CPP.Absyn.DFun p, Void arg) {
            // Check that name is not present already
            if (sig.get(p.id_) != null)
                throw new TypeException ("Function " + p.id_ + " already defined");

            // Add function to signature
            FunType ft = new FunType (p.type_, p.listarg_);
            sig.put(p.id_, ft);

            return null;
        }
    }

    // Type check a function definition.

    public class DefVisitor implements Def.Visitor<Void,Void> {
        public Void visit(CPP.Absyn.DFun p, Void arg) {
            // set return type and initial context
            returnType = p.type_;
            cxt = new LinkedList();
            cxt.add(new TreeMap());

            // add all function parameters to context
            for (Arg a: p.listarg_) {
                a.accept(new ArgVisitor(), arg);
            }

            // check function statements
            for (Stm s: p.liststm_) {
                s.accept(new StmVisitor(), arg);
            }

            return null;
        }
    }

    ///////////////////////// Function argument /////////////////////////

    // Add a type declaration to the context
    public class ArgVisitor implements Arg.Visitor<Void,Void> {
        public Void visit(CPP.Absyn.ADecl p, Void arg) {
            newVar (p.id_, p.type_);
            // // The variable should not be in the top block already.
            // if (cxt.get(0).containsKey(p.id_))
            //   throw new TypeException ("Local variable " + p.id_ + " already in current scope");
            // // Add it to the top block.
            // cxt.get(0).put (p.id_, p.type_);
            return null;
        }
    }

    ////////////////////////////// Statement //////////////////////////////

    public class StmVisitor implements Stm.Visitor<Void,Void> {
        public Void visit(CPP.Absyn.SExp p, Void arg) {
            Type t = p.exp_.accept(new ExpVisitor(), null);
            return null;
        }

        public Void visit(CPP.Absyn.SDecls p, Void arg) {
            if (p.type_ instanceof Type_void)
                throw new TypeException("Declarations can not have a void type");

            for (String id: p.listid_)
                newVar(id, p.type_);

            return null;
        }

        public Void visit(CPP.Absyn.SInit p, Void arg) {
            // p.id_ p.type_
            checkExpr(p.exp_, p.type_);
            newVar(p.id_, p.type_);
            return null;
        }

        public Void visit(CPP.Absyn.SReturn p, Void arg) {
            checkExpr(p.exp_, returnType);
            return null;
        }

        public Void visit(CPP.Absyn.SWhile p, Void arg) {
            checkExpr(p.exp_, BOOL);

            enterScope();
            p.stm_.accept(this, arg);
            leaveScope();

            return null;
        }

        public Void visit(CPP.Absyn.SBlock p, Void arg) {
            enterScope();
            for (Stm stm : p.liststm_)
                stm.accept(this, arg);
            leaveScope();
            return null;
        }

        public Void visit(CPP.Absyn.SIfElse p, Void arg) { 
            Type t = p.exp_.accept(new ExpVisitor(), arg);
            check(t, BOOL);
         enterScope();
         p.stm_1.accept(new StmVisitor(), arg);
         leaveScope();
         enterScope();
         p.stm_2.accept(new StmVisitor(), arg);
         leaveScope();
         return null;
        }
    }

    ////////////////////////////// Expression //////////////////////////////

    public class ExpVisitor implements Exp.Visitor<Type,Void> {

        // Literals
        public Type visit(CPP.Absyn.ETrue p, Void arg) {
            return BOOL;
        }
        
        public Type visit(CPP.Absyn.EFalse p, Void arg) {
            return BOOL;
        }

        public Type visit(CPP.Absyn.EInt p, Void arg) {
            return INT;
        }

        public Type visit(CPP.Absyn.EDouble p, Void arg) {
            return DOUBLE;
        }

        // Variable
        public Type visit(CPP.Absyn.EId p, Void arg) {
            return lookupVar(p.id_);
        }

        // Function call
        public Type visit(CPP.Absyn.EApp p, Void arg) {
            FunType ft = sig.get(p.id_);
            if (ft == null) throw new TypeException ("undefined function " + p.id_);
            if (ft.args.size() != p.listexp_.size())
                throw new TypeException ("wrong number of arguments");
            // check types of arguments
            int i=0;
            for (Exp e : p.listexp_) {
                ADecl a = (ADecl) ft.args.get(i);
                //checkExpr (e, a.type_);
                check(a.type_, e.accept(new ExpVisitor(), arg));
                i++;
            }
            return ft.returnType;
        }

        // Increment, decrement

        public Type visit(CPP.Absyn.EPostIncr p, Void arg) { 
            Type t = numericType(lookupVar(p.id_));
            return t;
        }

        public Type visit(CPP.Absyn.EPostDecr p, Void arg) { 
            Type t = numericType(lookupVar(p.id_));
            return t;
        }

        public Type visit(CPP.Absyn.EPreIncr p, Void arg) { 
            Type t = numericType(lookupVar(p.id_));
            return t;
        }

        public Type visit(CPP.Absyn.EPreDecr p, Void arg) { 
            Type t = numericType(lookupVar(p.id_));
            return t;
        }

        // Arithmetical operators

        public Type visit(CPP.Absyn.ETimes p, Void arg) { 
            Type t1 = p.exp_1.accept(new ExpVisitor(), arg);
            Type t2 = p.exp_2.accept(new ExpVisitor(), arg);
            numericType(t1);
            numericType(t2);
         check(t1, t2);
         return t1;
        }

        public Type visit(CPP.Absyn.EDiv p, Void arg) { 
            Type t1 = p.exp_1.accept(new ExpVisitor(), arg);
            Type t2 = p.exp_2.accept(new ExpVisitor(), arg);
         numericType(t1);
            numericType(t2);
         check(t1, t2);
         return t1;
        }

        public Type visit(CPP.Absyn.EPlus p, Void arg) { 
            Type t1 = p.exp_1.accept(new ExpVisitor(), arg);
            Type t2 = p.exp_2.accept(new ExpVisitor(), arg);
         numericType(t1);
            numericType(t2);
         check(t1, t2);
         return t1;
        }

        public Type visit(CPP.Absyn.EMinus p, Void arg) { 
            Type t1 = p.exp_1.accept(new ExpVisitor(), arg);
            Type t2 = p.exp_2.accept(new ExpVisitor(), arg);
         numericType(t1);
            numericType(t2);
         check(t1, t2);
         return t1;
        }

        // Comparison operators

        public Type visit(CPP.Absyn.ELt p, Void arg) { 
            Type t1 = p.exp_1.accept(new ExpVisitor(), null);
            Type t2 = p.exp_2.accept(new ExpVisitor(), null);
            numericType(t1);
            numericType(t2);
            equalTypes(t1, t2);
            return BOOL;
        }

        public Type visit(CPP.Absyn.EGt p, Void arg) { 
            Type t1 = p.exp_1.accept(new ExpVisitor(), null);
            Type t2 = p.exp_2.accept(new ExpVisitor(), null);
            numericType(t1);
            numericType(t2);
            equalTypes(t1, t2);
            return BOOL;
        }

        public Type visit(CPP.Absyn.ELtEq p, Void arg) { 
            Type t1 = p.exp_1.accept(new ExpVisitor(), null);
            Type t2 = p.exp_2.accept(new ExpVisitor(), null);
            numericType(t1);
            numericType(t2);
            equalTypes(t1, t2);
            return BOOL;
        }

        public Type visit(CPP.Absyn.EGtEq p, Void arg) { 
            Type t1 = p.exp_1.accept(new ExpVisitor(), null);
            Type t2 = p.exp_2.accept(new ExpVisitor(), null);
            numericType(t1);
            numericType(t2);
            equalTypes(t1, t2);
            return BOOL;
        }

        // Equality operators

        public Type visit(CPP.Absyn.EEq p, Void arg) { 
            Type t1 = p.exp_1.accept(new ExpVisitor(), arg);
            Type t2 = p.exp_2.accept(new ExpVisitor(), arg);
         equalTypes(t1, t2);
         check(t1, t2);
         return BOOL;
        }

        public Type visit(CPP.Absyn.ENEq p, Void arg) { 
            Type t1 = p.exp_1.accept(new ExpVisitor(), arg);
            Type t2 = p.exp_2.accept(new ExpVisitor(), arg);
         equalTypes(t1, t2);
         check(t1, t2);
         return BOOL;
        }

        // Logic operators

        public Type visit(CPP.Absyn.EAnd p, Void arg) { 
            Type t1 = p.exp_1.accept(new ExpVisitor(), arg);
            Type t2 = p.exp_2.accept(new ExpVisitor(), arg);
            check(t1, BOOL);
         check(t2, BOOL);
         return BOOL;
        }

        public Type visit(CPP.Absyn.EOr p, Void arg) { 
            Type t1 = p.exp_1.accept(new ExpVisitor(), arg);
            Type t2 = p.exp_2.accept(new ExpVisitor(), arg);
            check(t1, BOOL);
         check(t2, BOOL);
         return BOOL;
        }

        // Assignment
        public Type visit(CPP.Absyn.EAss p, Void arg) { 
            Type varType = lookupVar(p.id_);
            check(varType, p.exp_.accept(new ExpVisitor(), arg));
            return varType;
        }
    }

    ///////////////////////// Context handling /////////////////////////

    public void newVar(String x, Type t) {
        // Get the top context block
        Map<String,Type> m = cxt.get(0);
        // Check that the variable is not declared already
        if (m.containsKey(x))
            throw new TypeException("duplicate variable binding " + x);
        // Add the binding
        m.put(x, t);
    }

    public Type lookupVar(String x) {
        for (Map<String, Type> m: cxt) {
            Type t = m.get(x);
            if (t != null) return t;
        }
        throw new TypeException("Uninitialized variable " + x);
    }

    public void enterScope() { 
        cxt.add(0, new TreeMap());
    }

    public void leaveScope() { 
        cxt.remove(0);
    }

    ////////////////////////////// Exp / Type shape //////////////////////////////

    public void checkExpr(Exp e, Type t) {
        Type t1 = e.accept (new ExpVisitor(), null);
        check(t, t1);
    }

    public void check(Type t, Type u) {
        if (!t.equals(u)) {
            throw new TypeException("Expected type " + t + ", but found type " + u);
        }
    }

    public Type numericType(Type t) {
        if (!t.equals(INT) && !t.equals(DOUBLE)) {
            throw new TypeException("expected expression of numeric type");
        }

        return t;
    }

    public void equalTypes(Type t1, Type t2) {
      if (!t1.equals(t2))
         throw new TypeException("expected types " + t1 + " and " + t2 + " to be equal");
   }

    public Type booleanCheck(String operand, Type t1, Type t2) {
        if (t1 instanceof Type_bool && t2 instanceof Type_bool) {
            return BOOL;
        }
        else {
            throw new TypeException("The type must be boolean for " + operand);
        }
    }
}